### PR TITLE
Add saved questions picker

### PR DESCRIPTION
**Pull request by alxnddr**
### Description
Add saved question picker that shows collections tree.

#### Some important facts
- In custom mode when a table is selected and users join a saved question the picker shows saved questions only from the already selected database.
- When the saved question picker is active the search shows only saved questions.

### Screenshot
Screen Shot 2021-06-30 at 21 14 38

This is rework of the previous PR #15663 (which itself was the rework of PR #14864)

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/DataSelector.jsx [hunk 1]
```diff
@@ -705,12 +725,19 @@ export class UnconnectedDataSelector extends Component {
...
-    const { searchText } = this.state;
+    const { searchText, isSavedQuestionPickerShown } = this.state;
     const { canChangeDatabase, selectedDatabaseId } = this.props;
-    const searchDatabaseId = canChangeDatabase ? null : selectedDatabaseId;
+    const currentDatabaseId = canChangeDatabase ? null : selectedDatabaseId;
 
     const isSearchActive = searchText.trim().length >= MIN_SEARCH_LENGTH;
 
+    const searchPlaceholder = isSavedQuestionPickerShown
+      ? `Search for a question...`
```
Comment (ariya) [inline frontend/src/metabase/query_builder/components/DataSelector.jsx:735]: Just a note: I'm unsure if this adds a new string for the translator (late in the process), but it turns out that we have already a similar string "Search for a question" (without the triple dots). Either we use that (however, it'll be inconsistent with "Searching for a table...") or we use the generic "Search" (this string is already everywhere).
Alternatively, we just keep it as is (with the triple dots) and ask the translators to catch up in 40.1 or something.

Comment (mazameli) [reply frontend/src/metabase/query_builder/components/DataSelector.jsx]: If this is going in 0.40, we should use the already-translated string, please.

File: frontend/src/metabase/query_builder/components/data-search/utils.js [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
       ...searchResultItem,
-      id: `card__${searchResultItem.id}`,
+      id: `${SAVED_QUESTION_ID_PREFIX}${searchResultItem.id}`,
     };
   }
 
   return searchResultItem;
 }
+
+export function isSavedQuestion(table) {
```
Comment (ariya) [inline frontend/src/metabase/query_builder/components/data-search/utils.js:16]: Where is this function being used?

**Comment by ariya** [comment]
Overall, it looks super good :+1: .

There's a minor issue (likely because the state `isSavedQuestionPickerShown` isn't initialized properly) if an existing question is edited. Try New Custom Question, Saved Questions, choose any question and then Visualize. Then click the Show Editor to enter the notebook again and click on the question name. This would start over at Sample Dataset (should be showing saved question picker instead).

However, looks like it's only a minor annoyance and I'm fine if we address it as a follow-up (not necessarily lumped in this PR).

File: frontend/src/metabase/query_builder/components/DataSelector.jsx (hunk 1 above)
Comment (alxnddr) [reply frontend/src/metabase/query_builder/components/DataSelector.jsx]: I used the existing "Search for a question"

File: frontend/src/metabase/query_builder/components/data-search/utils.js (hunk 2 above)
Comment (alxnddr) [reply frontend/src/metabase/query_builder/components/data-search/utils.js]: Nice catch, removed it

**Comment by alxnddr** [comment]
@ariya thanks for catching this, I'll create a follow-up PR

File: frontend/src/metabase/query_builder/components/data-search/SearchResults.jsx [hunk 3]
```diff
@@ -9,10 +9,22 @@ import Search from "metabase/entities/search";
 
 import { SearchResultItem } from "./SearchResultItem";
 
-export function SearchResults({ searchQuery, onSelect, databaseId }) {
+const propTypes = {
+  databaseId: PropTypes.string,
+  searchQuery: PropTypes.string.required,
+  onSelect: PropTypes.func.required,
+  searchModels: PropTypes.arrayOf(PropTypes.string.isRequired),
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/data-search/SearchResults.jsx:16]: nip: should we maybe list the `searchModels` options right away if we know what's in use now?
Like `PropTypes.arrayOf(PropTypes.oneOf(["card", "table"]))`

File: frontend/src/metabase/query_builder/components/saved-question-picker/SavedQuestionPicker.jsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  databaseId: PropTypes.string,
+  tableId: PropTypes.string,
+};
+
+const OUR_ANALYTICS_COLLECTION = {
+  id: "root",
+  schemaName: "Everything else",
+  name: t`Our analytics`,
+  icon: "folder",
+};
```
Comment (kulyk) [inline frontend/src/metabase/query_builder/components/saved-question-picker/SavedQuestionPicker.jsx:37]: Should we maybe use the `ROOT_COLLECTION` object from [here](https://github.com/metabase/metabase/blob/ab50eaaa2fdf0583092f99ee11e8fc9a298582bf/frontend/src/metabase/entities/collections.js#L18), so we have a single source of truth on root ID and name?

**Review by ariya — APPROVED** [review]
Excellent work. Let's ship it :1st_place_medal:

**Review by kulyk — COMMENTED** [review]
Looks great, both the component and the code 👏

Noticed two a bit confusing behaviors. Should we address them or is it by design / out of scope of this PR?

### Collections tree sync

1. Let's say all my questions are in "Our analytics"
2. Move a question in any other collection, let's say to a personal one
3. Click "Ask a question" in the navigation bar > Simple Question > Saved Questions

**Expected**: the collection tree displays both "Our analytics" and my personal collection

**Actual**: the collection tree displays "Our analytics" only. However, the moved question is still there

**Note**: if you refresh the page, the tree is shown as expected

### Questions in nested collections

1. Let's say all my questions are in "Our analytics"
2. Create collections "C1" and "C2", "C2" has to be a child of "C1"
3. Move a question to "C2", "C1" has to be empty
4. Click "Ask a question" in the navigation bar > Simple Question
5. Refresh the page to avoid the previous issue
6. Select "Saved Questions"

**Expected**: there is either a "C2" in the collection tree or "C1" > "C2" tree

**Actual**: both "C1" and "C2" are not displayed, so the only way to access a question from "C2" is to use search

### Demo

I recorded a 2min video (with voiceover) where I reproduced both issues if my descriptions are unclear


https://user-images.githubusercontent.com/17258145/124159579-ff23e900-daa3-11eb-91ad-f789abd22038.mov

**Comment by ariya** [comment]
> **Note**: if you refresh the page, the tree is shown as expected

I believe this is a generic/known state sync issue (way before this PR). Try this: create a simple question and save it. Then start another question and try to find that new saved question. You won't immediately find it, until you refresh the page.

**Comment by alxnddr** [comment]
@kulyk thanks for your review! You caught the bug with hidden collections and I've fixed it now and addressed other suggestions. Regarding the tree sync without page reload I think we should address this problem separately because it works this way in the entire app

**Pull request merged**